File: utils/channelutils.py
from utils import globals
from utils import helpers

async def createChannels(member, origChannel):
	# fetch pre- channel creation data
	server = origChannel.guild
	savedchannelname = globals.data[f'creation vcs'][f'{origChannel.id}']
	channelname = '-'

	# create the vc
	created_vc = await server.create_voice_channel(channelname, category=origChannel.category, sync_permissions=True)

	# move creator to created channel
	await member.move_to(created_vc)

	# create the tc
	created_tc = await server.create_text_channel(channelname, category=origChannel.category, sync_permissions=False)

	# save data to file
	globals.data[f'created vcs'][f'{created_vc.id}'] = f'{created_tc.id}'
	globals.data[f'created tcs'][f'{created_tc.id}'] = f'{created_vc.id}'

	# set tc access permissions
	await created_tc.set_permissions(server.me, view_channel=True)
	await created_tc.set_permissions(server.default_role, view_channel=False)

	# send configuration message
	configMsg = await helpers.sendEmbed(created_tc, "Channel Configuration", f'react with 🔒 to make the voice channel private and with 🔓 to make it public again!\nCurrent status: unlocked', globals.defaultcolor)
	globals.data[f'cconfig msgs'][f'{created_tc.id}'] = f'{configMsg.id}'

	# set vc access permissions
	await created_vc.set_permissions(server.me, view_channel=True)
	await created_vc.set_permissions(server.default_role, view_channel=True)

	# get special name queues
	channelname = ''
	ignore = False
	for i in range(0, len(savedchannelname)):
		if not ignore:
			if savedchannelname[i] == f'\\':
				ignore = True
				continue
			if savedchannelname[i] == f'#':
				# TODO: implement per creation channel enumeration
				if not f'{origChannel.id}' in globals.data[f'channel enumerations']:
					globals.data[f'channel enumerations'][f'{origChannel.id}'] = {}

				continue
			if savedchannelname[i] == f'°':
				channelname += member.name
				continue
		channelname += savedchannelname[i]
		ignore = False

	# update channel names
	await created_vc.edit(name=channelname)
	await created_tc.edit(name=channelname)

	# add reactions to config msg
	await configMsg.add_reaction("🔒");
	await configMsg.add_reaction("🔓");

	# debug log
	print(f'User \"{member.name}\" created channel \"{channelname}\"')

async def deleteChannelsIfEmpty(vc):
	# if there is noone in the vc
	if len(vc.members) < 1:

		# get channel name
		channelname = vc.name

		# get tc by id
		tc = globals.bot.get_channel(int(globals.data[f'created vcs'][f'{vc.id}']))

		# delete channels
		await vc.delete()
		await tc.delete()

		# save data to file
		del globals.data[f'created vcs'][f'{vc.id}']
		del globals.data[f'created tcs'][f'{tc.id}']
		del globals.data[f'cconfig msgs'][f'{tc.id}']

		print(f'Channel \"{channelname}\" has been deleted')

async def setChannelAccess(user, vcID, viewChannel):
	# get channels
	tc = globals.bot.get_channel(int(globals.data[f'created vcs'][f'{vcID}']))

	# set permission
	await tc.set_permissions(user, view_channel=viewChannel)

File: utils/channelutils.js
async function createChannel(creationChannel, member) {
	// fetch name data
	let savedchannelname = data_creationVcs[`${creationChannel.id}`];
	let channelname = "-";

	// create vc
	let created_vc = await creationChannel.guild.channels.create(channelname, {
		type: 'voice',
		parent: creationChannel.parent
	});

	// move creator to channel
	member.voice.setChannel(created_vc);

	// create tc
	let created_tc = await creationChannel.guild.channels.create(channelname, {
		type: 'text',
		parent: creationChannel.parent
	});

	// save channel references
	data_createdVcs[created_vc] = created_tc;
	data_createdTcs[created_tc] = created_vc;

	// set tc permissions
	await created_tc.createOverwrite(client.user, {VIEW_CHANNEL: true});
	let tcperms = created_tc.permissionOverwrites;
	tcperms.forEach(tcpermow => {
		if (tcpermow.id !== client.user.id) tcpermow.delete();
	});
	await created_tc.createOverwrite(creationChannel.guild.roles.everyone, {VIEW_CHANNEL: false});
	created_tc.createOverwrite(member.user, {VIEW_CHANNEL: true});

	// assemble channel name
	channelname = '';
	ignore = false;
	for (var i = 0; i < savedchannelname.length; i++) {
		let cc = savedchannelname.charAt(i);
		if (!ignore) {
			if (cc === '\\') {
				ignore = true;
				continue;
			}
			else if (cc === '°') {
				channelname += member.user.username;
				continue;
			}
		}
		channelname += cc;
		ignore = false;
	}

	// update channel names
	created_vc.setName(channelname);
	created_tc.setName(channelname);

	// send configuration message
	let configMsg = await helpers.sendEmbed(created_tc, "Channel Configuration", "react with 🔒 to make the voice channel private and with 🔓 to make it public again!\nCurrent status: unlocked", defaultcolor);
	data_configMsgs[created_tc] = configMsg;

	// add reactions to config msg
	configMsg.react('🔒');
	configMsg.react('🔓');

	// log creation
	console.log(`User ${member.user.username} created Channel \"${channelname}\"`);
	
	// remove if empty
	removeChannelIfEmpty(created_vc);
};

async function removeChannelIfEmpty(channel) {
	if (channel.members.size < 1) {
		// get vc name
		let channelname = channel.name;

		// get tc
		tc = data_createdVcs[channel];

		// remove from lists
		delete data_configMsgs[tc];
		delete data_createdTcs[tc];
		delete data_createdVcs[channel];

		// delete channels
		channel.delete();
		tc.delete();

		// log channel removal
		console.log(`Channel \"${channelname}\" has been deleted`);
	}
};

async function addPerms(voiceChannel, member) {
	await data_createdVcs[voiceChannel].createOverwrite(member.user, {VIEW_CHANNEL: true});
};

async function removePerms(voiceChannel, member) {
	await data_createdVcs[voiceChannel].permissionOverwrites.get(member.id)?.delete();
};

// export functions
module.exports.createChannel = createChannel;
module.exports.removeChannelIfEmpty = removeChannelIfEmpty;
module.exports.addPerms = addPerms;
module.exports.removePerms = removePerms;
